crates/ruff_python_ast/src/str.rs

```diff
@@ -67,6 +69,15 @@ pub fn is_triple_quote(content: &str) -> bool {
     TRIPLE_QUOTE_STR_PREFIXES.contains(&content) || TRIPLE_QUOTE_BYTE_PREFIXES.contains(&content)
 }
 
+/// Return `true` if the string expression is an implicit concatenation.
+pub fn is_implicit_concatenation(content: &str) -> bool {
+    lexer::lex(content, Mode::Module)
+        .flatten()
+        .filter(|(_, tok, _)| matches!(tok, Tok::String { .. }))
+        .nth(1)
+        .is_some()
```

Comment:
This change regresses all our benchmarks by 3-5%. I assume that calling the lexer is the expensive operation. 

```
❯ cargo benchmark --baseline=main
    Finished bench [optimized] target(s) in 0.10s
     Running benches/linter.rs (target/release/deps/linter-b1b4f7dad7c3da18)
linter/default-rules/numpy/globals.py
                        time:   [114.58 µs 114.69 µs 114.81 µs]
                        thrpt:  [25.700 MiB/s 25.727 MiB/s 25.753 MiB/s]
                 change:
                        time:   [+3.4859% +3.6247% +3.7608%] (p = 0.00 < 0.05)
                        thrpt:  [-3.6245% -3.4979% -3.3685%]
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild
linter/default-rules/pydantic/types.py
                        time:   [2.2319 ms 2.2410 ms 2.2508 ms]
                        thrpt:  [11.331 MiB/s 11.380 MiB/s 11.427 MiB/s]
                 change:
                        time:   [+2.6956% +2.9809% +3.2544%] (p = 0.00 < 0.05)
                        thrpt:  [-3.1519% -2.8946% -2.6248%]
                        Performance has regressed.
Found 20 outliers among 100 measurements (20.00%)
  15 (15.00%) low severe
  3 (3.00%) high mild
  2 (2.00%) high severe
linter/default-rules/numpy/ctypeslib.py
                        time:   [1.0648 ms 1.0652 ms 1.0658 ms]
                        thrpt:  [15.623 MiB/s 15.632 MiB/s 15.638 MiB/s]
                 change:
                        time:   [+3.7858% +3.8922% +4.0261%] (p = 0.00 < 0.05)
                        thrpt:  [-3.8703% -3.7464% -3.6477%]
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
linter/default-rules/large/dataset.py
                        time:   [5.0569 ms 5.0704 ms 5.0813 ms]
                        thrpt:  [8.0064 MiB/s 8.0236 MiB/s 8.0450 MiB/s]
                 change:
                        time:   [+4.4705% +4.9010% +5.3636%] (p = 0.00 < 0.05)
                        thrpt:  [-5.0906% -4.6720% -4.2792%]
                        Performance has regressed.

linter/all-rules/numpy/globals.py
                        time:   [266.26 µs 266.46 µs 266.64 µs]
                        thrpt:  [11.066 MiB/s 11.074 MiB/s 11.082 MiB/s]
                 change:
                        time:   [+5.0624% +5.4244% +5.7740%] (p = 0.00 < 0.05)
                        thrpt:  [-5.4588% -5.1453% -4.8185%]
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
linter/all-rules/pydantic/types.py
                        time:   [3.6534 ms 3.6759 ms 3.6976 ms]
                        thrpt:  [6.8973 MiB/s 6.9380 MiB/s 6.9807 MiB/s]
                 change:
                        time:   [+0.0982% +0.5182% +0.9516%] (p = 0.02 < 0.05)
                        thrpt:  [-0.9427% -0.5156% -0.0981%]
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
linter/all-rules/numpy/ctypeslib.py
                        time:   [2.2603 ms 2.2640 ms 2.2688 ms]
                        thrpt:  [7.3392 MiB/s 7.3548 MiB/s 7.3667 MiB/s]
                 change:
                        time:   [+4.0998% +4.3497% +4.6664%] (p = 0.00 < 0.05)
                        thrpt:  [-4.4583% -4.1684% -3.9383%]
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe
linter/all-rules/large/dataset.py
                        time:   [8.6941 ms 8.6980 ms 8.7021 ms]
                        thrpt:  [4.6751 MiB/s 4.6773 MiB/s 4.6793 MiB/s]
                 change:
                        time:   [+1.8525% +2.0888% +2.3387%] (p = 0.00 < 0.05)
                        thrpt:  [-2.2853% -2.0461% -1.8188%]
                        Performance has regressed.
Found 25 outliers among 100 measurements (25.00%)
  13 (13.00%) low severe
  5 (5.00%) low mild
  3 (3.00%) high mild
  4 (4.00%) high severe
```

Can we apply a heuristic to short-circuit the call into the lexer? Or implement the test explicitly by:

* Testing if the start and end quotes are identical: if not -> implicit concatenation 
* find the closing quote character (ignoring escaped quotes): If it is at the end of contents -> all good, if not -> concatenation. We should be able to use `find` only for docstrings, if my assumption that triple quotes can't be escaped is correct

Reply:
Great call, horray for continuous benchmarking! Let me play around with this. I'm sure we can do better.

Reply:
Alright, let's see if this new version does any better...

Reply:
This looks much _better_ but there's still a noticeable difference in the NumPy case.

Reply:
Actually... is this just noise? The default-rules benchmark shouldn't be affected by this change.

Reply:
1-2% changes are most often just noise. Especially if they are specific to a single test.